src/net_processing.cpp

```diff
@@ -2683,7 +2684,7 @@ void ProcessMessage(
                     pfrom.fDisconnect = true;
                     return;
                 } else if (!fAlreadyHave && !chainman.ActiveChainstate().IsInitialBlockDownload()) {
-                    RequestTx(State(pfrom.GetId()), inv.hash, current_time);
+                    RequestTx(State(pfrom.GetId()), ToGenTxid(inv), current_time);
```

Comment:
`ToGenTxid()` expects that we constructed the `CInv` ourselves or that we verified that it's for a transaction type CInv. Here, the only guarantee that `inv` is transaction type is that AlreadyHave returns true for non-transaction and non-block type invs. That's not obvious and seems a bit brittle. Perhaps we should change this else branch into an `else if (inv.type == MSG_TX || inv.type == MSG_WITNESS_TX || inv.type == MSG_WTX)` to make this safer?

Reply:
I think it'd be better if `AlreadyHave` was split into different functions for block and tx, and the tx version took a GenTxid. Could be done in a follow up. I have a branch here that does that: https://github.com/jnewbery/bitcoin/tree/2020-07-split-already-have

Reply:
> Perhaps we should change this else branch into an `else if (inv.type == MSG_TX || inv.type == MSG_WITNESS_TX || inv.type == MSG_WTX)` to make this safer?

Agree, proposed in 66398c151374

Reply:
> `else if (inv.type == MSG_TX || inv.type == MSG_WITNESS_TX || inv.type == MSG_WTX)` to make this safer?

per change in `protocol.h:L447` this would now be simpler: `else if inv.IsGenTxMsg()`

Reply:
Let's fix that by splitting AlreadyHave up separately.

Reply:
Done in #19610.